graal-js/src/com.oracle.truffle.js/src/com/oracle/truffle/js/builtins/SetPrototypeBuiltins.java

```diff
@@ -229,6 +264,406 @@ protected boolean hasNoObject(Object thisObj, Object key) {
+        }
+
+        protected Object addEntryFromIterable(Object target, Object iterable, Object adder) {
+            if (!JSRuntime.isCallable(adder)) {
+                adderError.enter();
+                throw Errors.createTypeErrorCallableExpected();
+            }
+            IteratorRecord iteratorRecord = getIteratorNode.execute(iterable);
+            try {
+                while (true) {
+                    Object next = iteratorStepNode.execute(iteratorRecord);
+                    if (next == Boolean.FALSE)
+                        return target;
+                    Object nextValue = iteratorValueNode.execute((DynamicObject) next);
+                    call(adder, target, nextValue);
+                }
+            } catch (Exception ex) {
+                iteratorError.enter();
+                iteratorCloseAbrupt(iteratorRecord.getIterator());
+                throw ex;
+            }
+        }
+
+        protected final void iteratorCloseAbrupt(DynamicObject iterator) {
+            if (iteratorCloseNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                iteratorCloseNode = insert(IteratorCloseNode.create(getContext()));
+            }
+            iteratorCloseNode.executeAbrupt(iterator);
+        }
+
+        protected Object call(Object function, Object target, Object... userArguments) {
+            if (callFunctionNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                callFunctionNode = insert(JSFunctionCallNode.createCall());
+            }
+            return callFunctionNode.executeCall(JSArguments.create(target, function, userArguments));
+        }
+
+        protected final Object getAddFunction(Object object) {
+            if (getAddNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                getAddNode = insert(PropertyGetNode.create("add", false, getContext()));
+            }
+            return getAddNode.getValue(object);
+        }
+
+        protected final Object getRemoveFunction(Object object) {
+            if (getRemoveNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                getRemoveNode = insert(PropertyGetNode.create("delete", false, getContext()));
+            }
+            return getRemoveNode.getValue(object);
+        }
+
+        protected final Object getHasFunction(Object object) {
+            if (getHasNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                getHasNode = insert(PropertyGetNode.create("has", false, getContext()));
+            }
+            return getHasNode.getValue(object);
+        }
+
+        protected final Object constructSet(Object... argeuments) {
+            Object ctr = getContext().getRealm().getSetConstructor();
+            return JSRuntime.construct(ctr, argeuments);
```

Comment:
This can probably be improved as well, but I am fine with leaving it as-is for the moment.